### ISSUE TITLE
Key masking plan

### ISSUE THREAD

**Issue by gedankenexperimenter** [gedankenlab/Kaleidoglyph-Unshifter#2]
During the `eventHandlerHook()` pass, don't put any keycodes into the HID report for keys that are in our dictionary.

Then, during the `startLoopHook()` pass (or pre-clear `loopHook` if the plugin interface doesn't get updated), do the following steps:

1. Check the modifier state.
2. Look up keycodes for all `Unshift`'d keys (based on modifier state).
3. Add keycodes for any `Unshift`'d keys that don't want to change the modifier state.
4. If _all_ `Unshift`'d keys want to change the modifier state, _and_ the HID report is empty (except for modifiers), change modifier state and add keycodes for `Unshift`'d keys.

This method still could cause problems if the following sequence takes place rapidly:

1. press `unshift`'d key
2. press normal key
3. release normal key
4. release `unshift`'d key

In this case, you'd get a spurious `unshift`'d character. It would be better if we could, on press of an `unshift`'d key that wants to toggle the shift state, mask all the currently held normal keys, so have a "most recent keypress wins" policy, but that would require a lot more bookkeeping.

**Comment by gedankenexperimenter** [comment]
Now that I've implemented gedankenlab/Qukeys, I think this is not going to work well enough. Whenever there are keys in the Qukeys queue, we need to suppress them here as well to keep them from getting inserted out of order (A quick `shift`+`X` would produce the wrong symbol, depending on `X`).

**Comment by gedankenexperimenter** [comment]
I need to figure out how key masking works in Kaleidoscope. Maybe that will solve the problem, because if a plugin (e.g. Qukeys) wants to suppress a key's input, it can call `maskKey()`, and other plugins (e.g. Unshifter) can know about that and act accordingly. (and when Qukeys pops a key from its queue, it would call `unmaskKey()`)

**Comment by gedankenexperimenter** [comment]
Masking _keys_ won't be practical; we would have to make guesses or do way too much bookkeeping in order to figure out which key(s) to mask. What we can do instead is mask _keycodes_.

How does this work?  Not worrying about other plugins for the moment, we use the `eventHandlerHook` to identify unshift'd keys. At that point, we don't know which keycode to use for them, however. Later, in the pre-report hook, we check the key report (now complete except for unshift'd keys), look at the shift state, and then make a list of `Key`s that should be added (they should be stored in a queue when toggled on, à la Qukeys). If none of them need to toggle the shift state, just go ahead and add them (probably in a series of HID reports, one for each key toggled on -- realistically we don't expect more than one). If a key does need the shift state toggled, do that, but build a keycode mask from the existing HID report, and remove those codes from the report, until we see a HID report without them (each one is independent).

Basically, this means that when I press a key that flips the shift state, it cancels any held key(codes) and suppresses them until those keys are released.

I _think_ it's possible to do something similar with INJECTED key events during the key scan, but I'm not sure.

This is a lot of trouble for what seems like it's probably a corner case.

**Comment by gedankenexperimenter** [comment]
I've decided not to implement this plugin in the current Kaleidoscope; I have much better ideas for how it will work in my rewrite.
